### ISSUE TITLE
Need to access component template that is within an external module

### ISSUE THREAD

**Issue by JSilva94** [isaacplmann/ngx-tour#51]
I have a tourAnchor in component A that is in Module A. Module A is in AppModule, and so is my component that has my tourService. However, I cannot access the tourAnchor in component A. I can only directly access tourAnchors in components that are directly imported into the appModule. I have tried to export component A in module A, and nothing happens. 

Any suggestions?

**Comment by isaacplmann** [comment]
The tourService needs to be a singleton.  TourAnchors automatically register themselves with the tourService, but if they're registering themselves with a tourService that is different from the top level tour service then the top level tour service won't know they exist.

Does that help?

**Comment by JSilva94** [comment]
I believe my tourService is a singleton. The problem is that my my tourService cannot find the tourAnchor in componentA as it is not directly imported into the same module. 

I dont really understand why

**Comment by isaacplmann** [comment]
The tourService doesn't try to find any tourAnchors.  The tourAnchors register themselves with the tourService.  The fact that your app is compiling and running means that the tourAnchor has found some tourService to register itself with.  But the fact the app component level tourService doesn't know about the tourAnchor means that the tourService that the tourAnchor registered itself with is not the top level one.

If module A is a lazy loaded module (using `loadChildren: 'somestring'` in the route) then the providers in that module are different from the top level module.  To work around this, have both the AppModule and ModuleA import a SharedModule which takes care of importing the TourModule.

**Comment by JSilva94** [comment]
I am trying that and it is still not finding it. Do I need to export my providers in the tourModule?
I have a tourComponent that does the initializing. This component is declared in my tourModule. 

Thank you for your help

**Comment by JSilva94** [comment]
TutorialComponent calls initialize

TutorialModule declares TutorialComponent and exports it.

headerComponent.html has the tutorialComponent Selector in it.

LayoutModule has tutorialModule imported, and headerComponent declared

The tutorial button shows up, and it also console logs all the anchors that are initialized.

However the tutorial ends immediately after click. I assume this is because it cannt find it

**Comment by JSilva94** [comment]
basically, it ends when the tourAnchor is in a <router-outlet> and the component that is generated from the <router-outlet> is lazyloaded

**Comment by isaacplmann** [comment]
Where are you importing `TourMdMenuModule.forRoot()` or `TourNgBootstrapModule.forRoot()`?

**Comment by JSilva94** [comment]
in my layoutModule

**Comment by JSilva94** [comment]
it is able to find components that are directly imported into the layout module, but it cannot find componentA that is imported into moduleA. Howveer ModuleA is imported into LayoutMOdule

**Comment by isaacplmann** [comment]
Can you double check that you're not importing `TourMdMenuModule.forRoot()` or providing the `TourService` in ModuleA?

If that's not the issue I'll need some minimal reproduction repo to help track this down.

**Comment by JSilva94** [comment]
Nope, I am not doing that.

Where am I suppose to initialize the tourService?

**Comment by isaacplmann** [comment]
You can initialize it anywhere.  It just needs to happen before someone tries to start the tour - or the tour won't start.

**Comment by JSilva94** [comment]
okay, so in layoutComponent I initialize the tour. I then have a layout-router.module that lazyloads modules. Any of those modules that have a touranchor cause the tour to end when it gets to that step and I do not know why

**Comment by isaacplmann** [comment]
Are you specifying a route in those tour steps?  If the tourAnchor is not loaded on the screen, the tour won't be able to find it.  But if you specify a route in the step, the tourService will navigate to that route before trying to open that tour step.

**Comment by JSilva94** [comment]
The tour does find it because when it logs the events, it logs the initializer step and it shows me all the objects that it initialized.

**Comment by JSilva94** [comment]
I would like to thank you for all your help.

I have found the problem. Any component that is not declared in the same module as TourNgBootstrapModule, then it will not be found.

**Comment by JSilva94** [comment]
If i declare componentB in ModuleA, the tour works. However, If I export componentB in Module B and then import ModuleB into Component A it does not work. What am i doing wrong?

The tourNgboostrapmodule.forRoot() is in moduleA

**Comment by isaacplmann** [comment]
I don't understand your descriptions.  I need to see actual code to help further.

**Comment by kahboom** [comment]
Apologies in advance for the long comment. I'm running into a similar issue. Basically, what is the easiest way to share the same instance of TourNgxBootstrapModule in a parent module (say `AppModule`) with its own `AppComponent` that lazily loads other Angular child modules, which have their own child components.

This works: You import ngx-tour module into the parent module and instantiate the tour service from the parent component, even if the routes it references are defined in a lazily-loaded child module (as long as anchors are set up properly in the child component's respective HTML). This is similar to the "other-route" example in the demo.

The downside is that for this to work the steps need to be defined in the parent component tour service when you first initialize the tour service ([this group](https://github.com/isaacplmann/ngx-tour/blob/master/src/app/ngx-bootstrap/ngx-bootstrap.component.ts#L11)). There are two problems with this:

1. You need to use Prev/Next for this to work, as it uses the route you defined in the step in that parent component (like [this](https://github.com/isaacplmann/ngx-tour/blob/master/src/app/ngx-bootstrap/ngx-bootstrap.component.ts#L40)). It will redirect the user to that step's route. Sometimes we want to instantiate the service if the user just navigates to that route on their own, whether it be instantiated via the component or the route. Maybe we don't want to force a specific flow with Prev/Next.
2. You cannot group/separate out steps unless they are in components that belongs to the same parent module. If you try to import the tour module separately into a lazily-loaded child module, it will throw an error stating that it has already been imported by the parent module. If you just use the service, it doesn't _seem_ to detect the anchor, maybe because it is deeply nested in a lazily-loaded child module component's respective HTML (woah, that's a mouth full). Will test further.

Here is what I mean:

```
<!-- Parent module & its components -->
<router-outlet>
  <!-- Lazily-loaded child module & its components-->
  <router-outlet>
    <p tourAnchor="tooDeep"></p>
  </router-outlet>
</router-outlet>
```

**tl;dr: 1) How can we group steps between modules that are lazily loaded with a common parent module? Notably between lazily-loaded child modules and a common parent. This might also allow us to: 2) Not force the user through a flow with Prev/Next, but instantiate the tour service with a separate group of steps that are specific to that page (which is perhaps from a lazily-loaded child module component).**

A few use cases for this:
- Common parts, like a navigation bar or sidebar, might need steps (e.g. "Logout from here" in the navigation bar), and then lazily-loaded child module components might have their own steps (e.g. "This content means blah blah").
- Trigger the tour service for a subset of steps that are specific to the route, even if it's lazily loaded. For instance, a group of steps that should only show Prev/Next for that specific component. It's essentially like a mini guided tour within a parent guided tour. It sounds like a bad idea, but I think with nested modules this flexibility could be necessary.

I hope it's clear. Thanks!

**Comment by isaacplmann** [comment]
@kahboom I would do something like this:

```ts
// app.module.ts
@Module({
  import: [ TourNgxBootstrapModule.forRoot() ],
  // ...
})
export class AppModule {}

// lazyLoadedOne.ts
ngOnInit() {
  this.tourService.initialize([ /* custom steps for this component */ ]);
}

// lazyLoadedTwo.ts
ngOnInit() {
  this.tourService.initialize([ /* custom steps for this component */ ]);
}
```

If you wanted to have global tour steps that aren't overwritten by the individual lazy loaded components, you could do this:

```ts
// globalTourSteps.service.ts
export const globalTourSteps = [
  // steps here
];

// lazyLoadedOne.ts
ngOnInit() {
  this.tourService.initialize([ ...globalTourSteps, /* custom steps for this component */ ]);
}
```

Does that accomplish what you're trying to do?

**Comment by kahboom** [comment]
@isaacplmann - Thanks for the suggestion. Hmm, it's not working 100% for me, but I will try to put together a plunkr in a bit and keep testing a few things. For one of the routes it doesn't seem to trigger the step.

**Comment by kahboom** [comment]
Sorry for the delay, have another demo this Friday. I think I got it--the issue is when the tourAnchors are defined in the templates of the lazily-loaded module components. So, you can initialize the tourService from any of those components, and if `tourAnchor` is in the HTML from the parent module component template it will work, but if the `tourAnchor` is defined in one of those nested templates it won't detect it. I tried initializing from `ngAfterViewInit` but that doesn't seem to help, maybe related to `ViewEncapsulation`.

**Comment by kahboom** [comment]
Okay, I'm a step closer. I had to declare the `TourNgxBootstrapModule` as an `import` and `TourService` as a `provider` in the lazily-loaded module. However, it seems to get appended to the outer view that contains the HTML from the child component, which is very annoying. See video below.

Also need to figure out a way to disable the tour globally once a user dismisses it. Will report back once I have more details. Perhaps I should have opened a separate issue 😂 . I can do that if you prefer @isaacplmann !

https://www.useloom.com/share/1cce9f90a36f465f950485a8de523670

**Comment by isaacplmann** [comment]
Ok. Great.  I just made a fix to the plunkr: https://plnkr.co/edit/8XjSDhOHiOewpOmAAV1P?p=preview

**Comment by kahboom** [comment]
Thanks @isaacplmann - everything is okay now except that the 'close' button only works when the anchor is loaded from the parent module template. Checking for any errors, had a colleague review and we're out of ideas (and time! lol). I'd imagine I'd run into a similar issue with prev/next but haven't needed it yet as the tour is isolated per page for the purposes of our project.

I'm trying to update the Plunker to better reflect the code: https://plnkr.co/edit/9HOvtJ?p=streamer however the content is not showing up for some reason.

This is my PR if you can make sense of it (no worries, I don't expect you to dig through all of that, but in case you are bored lol): https://github.com/syndesisio/syndesis-ui/pull/1157/files

The tour that uses `tourAnchor` in the parent template seems to work without issues (in my PR it's the 'settings' component; the  troublesome ones are under 'connections' directory and its child components, like 'connections/create-page' or 'connections/create-page/connection-basics'.

I did determine I didn't need to inject the ngx-tour module into the lazily-loaded modules, but that I needed to at least list the tour service as a provider in the module or component.

**Comment by isaacplmann** [comment]
I think it's the fact that you're providing the `TourService` in the lazy loaded module that is causing the problem.  If you provide the `TourService` in a lazy loaded module, it creates a separate instance of the `TourService` which might not be the same one that other parts of your app are using.  I recommend `import`ing `NgxTourNgxBootstrap.forRoot()` in your `AppModule` and `import`ing `NgxTourNgxBootstrap` (without the `forRoot()`) in your lazy loaded modules.

**Comment by mikecabana** [comment]
I'll give it a go later on tonight when I'm home

**Comment by kahboom** [comment]
Thanks @mikecabana !
